servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java

```diff
@@ -572,7 +574,18 @@ public void closeGracefully() throws Exception {
                         @Override
                         public void handle(final GrpcServiceContext ctx, final BlockingIterable<Req> request,
                                            final GrpcPayloadWriter<Resp> responseWriter) throws Exception {
-                            route.handle(ctx, requireNonNull(request.iterator().next()), responseWriter);
+                            final Req firstItem;
+                            try (BlockingIterator<Req> requestIterator = request.iterator()) {
+                                firstItem = requireNonNull(requestIterator.next(), "Request item is null");
+                                if (requestIterator.hasNext()) {
+                                    // Consume the next item to make sure it's not a TerminalNotification with an error
+                                    requestIterator.next();
+                                    throw new GrpcStatus(INVALID_ARGUMENT, null,
+                                            "Only a single request item is expected, but saw the second one")
```

Comment:
Not changed in this PR but as you are changing code; this would be a better message.

```suggestion
                                            "More than one request message received.")
```